### PR TITLE
fix: Loosen iminuit tolerances for optimization tests

### LINKED ISSUE #1305: iminuit v2.4.0 breaks test_optim
# Description

[`iminuit` `v2.4.0`](https://github.com/scikit-hep/iminuit/releases/tag/v2.4.0) (released today 2021-02-10) is breaking the tests for 32b and 64b minuit in the `test_minimize` tests

https://github.com/scikit-hep/pyhf/blob/bc32a695aaea0b79b3f0b40676446452d115cf8e/tests/test_optim.py#L49

```pytb
_______________ test_minimize[do_grad-minuit-jax-64b-do_stitch] ________________

tensorlib = <class 'pyhf.tensor.jax_backend.jax_backend'>, precision = '64b'
optimizer = <class 'pyhf.optimize.minuit_optimizer'>, do_grad = True
do_stitch = True

    @pytest.mark.parametrize('do_stitch', [False, True], ids=['no_stitch', 'do_stitch'])
    @pytest.mark.parametrize('precision', ['32b', '64b'], ids=['32b', '64b'])
    @pytest.mark.parametrize(
        'tensorlib',
        [
            pyhf.tensor.numpy_backend,
            pyhf.tensor.pytorch_backend,
            pyhf.tensor.tensorflow_backend,
            pyhf.tensor.jax_backend,
        ],
        ids=['numpy', 'pytorch', 'tensorflow', 'jax'],
    )
    @pytest.mark.parametrize(
        'optimizer',
        [pyhf.optimize.scipy_optimizer, pyhf.optimize.minuit_optimizer],
        ids=['scipy', 'minuit'],
    )
    @pytest.mark.parametrize('do_grad', [False, True], ids=['no_grad', 'do_grad'])
    def test_minimize(tensorlib, precision, optimizer, do_grad, do_stitch):
        pyhf.set_backend(tensorlib(precision=precision), optimizer())
        m = pyhf.simplemodels.hepdata_like([50.0], [100.0], [10.0])
        data = pyhf.tensorlib.astensor([125.0] + m.config.auxdata)
        # numpy does not support grad
        if pyhf.tensorlib.name == 'numpy' and do_grad:
            with pytest.raises(pyhf.exceptions.Unsupported):
                pyhf.infer.mle.fit(data, m, do_grad=do_grad)
        else:
            identifier = f'{"do_grad" if do_grad else "no_grad"}-{pyhf.optimizer.name}-{pyhf.tensorlib.name}-{pyhf.tensorlib.precision}'
            expected = {
                # numpy does not do grad
                'do_grad-scipy-numpy-32b': None,
                'do_grad-scipy-numpy-64b': None,
                'do_grad-minuit-numpy-32b': None,
                'do_grad-minuit-numpy-64b': None,
                # no grad, scipy, 32b - never works
                'no_grad-scipy-numpy-32b': [1.0, 1.0],
                'no_grad-scipy-pytorch-32b': [1.0, 1.0],
                'no_grad-scipy-tensorflow-32b': [1.0, 1.0],
                'no_grad-scipy-jax-32b': [1.0, 1.0],
                # no grad, scipy, 64b
                'no_grad-scipy-numpy-64b': [0.49998815367220306, 0.9999696999038924],
                'no_grad-scipy-pytorch-64b': [0.49998815367220306, 0.9999696999038924],
                'no_grad-scipy-tensorflow-64b': [0.49998865164653106, 0.9999696533705097],
                'no_grad-scipy-jax-64b': [0.4999880886490433, 0.9999696971774877],
                # do grad, scipy, 32b
                'do_grad-scipy-pytorch-32b': [0.49993881583213806, 1.0001085996627808],
                'do_grad-scipy-tensorflow-32b': [0.4999384582042694, 1.0001084804534912],
                'do_grad-scipy-jax-32b': [0.4999389052391052, 1.0001085996627808],
                # do grad, scipy, 64b
                'do_grad-scipy-pytorch-64b': [0.49998837853531425, 0.9999696648069287],
                'do_grad-scipy-tensorflow-64b': [0.4999883785353142, 0.9999696648069278],
                'do_grad-scipy-jax-64b': [0.49998837853531414, 0.9999696648069285],
                # no grad, minuit, 32b - not very consistent for pytorch
                'no_grad-minuit-numpy-32b': [0.49622172117233276, 1.0007264614105225],
                #    nb: macos gives different numerics than CI
                # 'no_grad-minuit-pytorch-32b': [0.7465415000915527, 0.8796938061714172],
                'no_grad-minuit-pytorch-32b': [0.9684963226318359, 0.9171305894851685],
                'no_grad-minuit-tensorflow-32b': [0.5284154415130615, 0.9911751747131348],
                # 'no_grad-minuit-jax-32b': [0.5144518613815308, 0.9927923679351807],
                'no_grad-minuit-jax-32b': [0.49620240926742554, 1.0018986463546753],
                # no grad, minuit, 64b - quite consistent
                'no_grad-minuit-numpy-64b': [0.5000493563629738, 1.0000043833598724],
                'no_grad-minuit-pytorch-64b': [0.5000493563758468, 1.0000043833508256],
                'no_grad-minuit-tensorflow-64b': [0.5000493563645547, 1.0000043833598657],
                'no_grad-minuit-jax-64b': [0.5000493563528641, 1.0000043833614634],
                # do grad, minuit, 32b
                'do_grad-minuit-pytorch-32b': [0.5017611384391785, 0.9997190237045288],
                'do_grad-minuit-tensorflow-32b': [0.5012885928153992, 1.0000673532485962],
                # 'do_grad-minuit-jax-32b': [0.5029529333114624, 0.9991086721420288],
                'do_grad-minuit-jax-32b': [0.5007095336914062, 0.9999282360076904],
                # do grad, minuit, 64b
                'do_grad-minuit-pytorch-64b': [0.500273961181471, 0.9996310135736226],
                'do_grad-minuit-tensorflow-64b': [0.500273961167223, 0.9996310135864218],
                'do_grad-minuit-jax-64b': [0.5002739611532436, 0.9996310135970794],
            }[identifier]
    
            result = pyhf.infer.mle.fit(data, m, do_grad=do_grad, do_stitch=do_stitch)
    
            rtol = 2e-06
            # handle cases where macos and ubuntu provide very different results numerical
            if 'no_grad-minuit-tensorflow-32b' in identifier:
                # not a very large difference, so we bump the relative difference down
                rtol = 3e-02
            if 'no_grad-minuit-pytorch-32b' in identifier:
                # quite a large difference
                rtol = 3e-01
            if 'do_grad-minuit-pytorch-32b' in identifier:
                # a small difference
                rtol = 7e-05
            if 'no_grad-minuit-jax-32b' in identifier:
                rtol = 4e-02
            if 'do_grad-minuit-jax-32b' in identifier:
                rtol = 5e-03
    
            # check fitted parameters
>           assert pytest.approx(expected, rel=rtol) == pyhf.tensorlib.tolist(
                result
            ), f"{identifier} = {pyhf.tensorlib.tolist(result)}"
E           AssertionError: do_grad-minuit-jax-64b = [0.500049321731032, 1.0000044174002167]
E           assert approx([0.5002739611532436 ± 1.0e-06, 0.9996310135970794 ± 2.0e-06]) == [0.500049321731032, 1.0000044174002167]
E            +  where approx([0.5002739611532436 ± 1.0e-06, 0.9996310135970794 ± 2.0e-06]) = <function approx at 0x7fb30c6b6e50>([0.5002739611532436, 0.9996310135970794], rel=2e-06)
E            +    where <function approx at 0x7fb30c6b6e50> = pytest.approx
E            +  and   [0.500049321731032, 1.0000044174002167] = <bound method jax_backend.tolist of <pyhf.tensor.jax_backend.jax_backend object at 0x7fb210064b00>>(DeviceArray([0.50004932, 1.00000442], dtype=float64))
E            +    where <bound method jax_backend.tolist of <pyhf.tensor.jax_backend.jax_backend object at 0x7fb210064b00>> = <pyhf.tensor.jax_backend.jax_backend object at 0x7fb210064b00>.tolist
E            +      where <pyhf.tensor.jax_backend.jax_backend object at 0x7fb210064b00> = pyhf.tensorlib

tests/test_optim.py:126: AssertionError
```

`test_minuit_strategy_do_grad`

https://github.com/scikit-hep/pyhf/blob/bc32a695aaea0b79b3f0b40676446452d115cf8e/tests/test_optim.py#L182

as well as in `test_minuit_strategy_global` tests

https://github.com/scikit-hep/pyhf/blob/bc32a695aaea0b79b3f0b40676446452d115cf8e/tests/test_optim.py#L210

```pytb
__________________ test_minuit_strategy_global[tensorflow-1] ___________________

self = <pyhf.optimize.minuit_optimizer object at 0x7fb2107be700>
func = <function wrap_objective.<locals>.func at 0x7fb228255a60>
x0 = [1.0, 1.0], do_grad = True, bounds = [(0, 10), (1e-10, 10.0)]
fixed_vals = [], options = {}

    def _internal_minimize(
        self, func, x0, do_grad=False, bounds=None, fixed_vals=None, options={}
    ):
    
        minimizer = self._get_minimizer(
            func, x0, bounds, fixed_vals=fixed_vals, do_grad=do_grad
        )
        result = self._minimize(
            minimizer,
            func,
            x0,
            do_grad=do_grad,
            bounds=bounds,
            fixed_vals=fixed_vals,
            options=options,
        )
    
        try:
>           assert result.success
E           AssertionError

src/pyhf/optimize/mixins.py:49: AssertionError

During handling of the above exception, another exception occurred:

mocker = <pytest_mock.plugin.MockerFixture object at 0x7fb1e3d59370>
backend = (<pyhf.tensor.tensorflow_backend.tensorflow_backend object at 0x7fb2433ed080>, None)
strategy = 1

    @pytest.mark.parametrize('strategy', [0, 1])
    def test_minuit_strategy_global(mocker, backend, strategy):
        pyhf.set_backend(pyhf.tensorlib, pyhf.optimize.minuit_optimizer(strategy=strategy))
        spy = mocker.spy(pyhf.optimize.minuit_optimizer, '_minimize')
        m = pyhf.simplemodels.hepdata_like([50.0], [100.0], [10.0])
        data = pyhf.tensorlib.astensor([125.0] + m.config.auxdata)
    
        do_grad = pyhf.tensorlib.default_do_grad
>       pyhf.infer.mle.fit(data, m)

tests/test_optim.py:217: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/pyhf/infer/mle.py:122: in fit
    return opt.minimize(
src/pyhf/optimize/mixins.py:157: in minimize
    result = self._internal_minimize(**minimizer_kwargs, options=kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyhf.optimize.minuit_optimizer object at 0x7fb2107be700>
func = <function wrap_objective.<locals>.func at 0x7fb228255a60>
x0 = [1.0, 1.0], do_grad = True, bounds = [(0, 10), (1e-10, 10.0)]
fixed_vals = [], options = {}

    def _internal_minimize(
        self, func, x0, do_grad=False, bounds=None, fixed_vals=None, options={}
    ):
    
        minimizer = self._get_minimizer(
            func, x0, bounds, fixed_vals=fixed_vals, do_grad=do_grad
        )
        result = self._minimize(
            minimizer,
            func,
            x0,
            do_grad=do_grad,
            bounds=bounds,
            fixed_vals=fixed_vals,
            options=options,
        )
    
        try:
            assert result.success
        except AssertionError:
            log.error(result, exc_info=True)
>           raise exceptions.FailedMinimization(result)
E           pyhf.exceptions.FailedMinimization: Optimization failed. Estimated distance to minimum too large.
```

We will need to investigate what's up and perhaps loosen tolerances for `iminuit`.

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #1305 

Update the minimum required release of `iminuit` in the `minuit` extra to `v2.4.0`. This changes some of the minimization values which requires loosening of `iminuit`'s tolerances to reach convergence.

As [Hans notes](https://github.com/scikit-hep/pyhf/issues/1305#issuecomment-777223845):

> Remember that Minuit's tolerance and convergence criterion stops the iteration earlier than other minimisers, once the change in the parameter values becomes negligible compared to the parameter errors. This makes sense for statistical fits. Small means 1e-2 to 1e-3 of the error.

~~While this PR does introduce some large changes in the tests it also seems that `iminuit` `v2.4.0` gives more consistent results across OS and across CI vs. local.~~

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Set minimum compatible release of iminuit to v2.4.0 in 'minuit' extra
* Update expected values for iminuit minimizations in test/test_optim.py
* Loosen tolerances for iminuit convergence in test/test_optim.py
   - Needed for TensorFlow, but applied universally
```

### CODE REVIEW

File: tests/test_optim.py [hunk 1]
```diff
@@ -83,47 +83,47 @@ def test_minimize(tensorlib, precision, optimizer, do_grad, do_stitch):
...
             'no_grad-minuit-jax-64b': [0.5000493563528641, 1.0000043833614634],
             # do grad, minuit, 32b
-            'do_grad-minuit-pytorch-32b': [0.5017611384391785, 0.9997190237045288],
-            'do_grad-minuit-tensorflow-32b': [0.5012885928153992, 1.0000673532485962],
-            # 'do_grad-minuit-jax-32b': [0.5029529333114624, 0.9991086721420288],
-            'do_grad-minuit-jax-32b': [0.5007095336914062, 0.9999282360076904],
+            # large divergence by tensorflow and pytorch
+            'do_grad-minuit-pytorch-32b': [0.9731879234313965, 0.9999999403953552],
+            'do_grad-minuit-tensorflow-32b': [0.9366918206214905, 0.9126002788543701],
+            'do_grad-minuit-jax-32b': [0.5003563165664673, 0.9998618364334106],
```
Comment (matthewfeickert) [inline tests/test_optim.py:98]: It isn't really clear to me what is happenng that there is such a drastic shift here.

Comment (alexander-held) [reply tests/test_optim.py:98]: Could it be that the PyTorch/TensorFlow fail to converge?

Comment (matthewfeickert) [reply tests/test_optim.py:98]: @alexander-held That's a good question, but the are succeeding 

```python
import pyhf

precision = "32b"
do_grad = True
do_stitch = True
strategy = None
# strategy = 2
verbose = 1

names = ["TensorFlow", "PyTorch", "JAX"]
backends = [
    pyhf.tensor.tensorflow_backend,
    pyhf.tensor.pytorch_backend,
    pyhf.tensor.jax_backend,
]
for name, backend in zip(names, backends):
    print(f"\n{name}\n")
    pyhf.set_backend(
        backend(precision=precision),
        pyhf.optimize.minuit_optimizer(verbose=verbose, strategy=strategy),
    )
    model = pyhf.simplemodels.hepdata_like([50.0], [100.0], [10.0])
    data = pyhf.tensorlib.astensor([125.0] + model.config.auxdata)
    result = pyhf.infer.mle.fit(
        data, model, do_grad=do_grad, do_stitch=do_stitch, return_result_obj=True
    )
    fit_params = result[0]
    fit_result = result[1]
    print(f"fit params: {[pyhf.tensorlib.tolist(x) for x in fit_params]}")
    print(fit_result.minuit)
    print(f"success: {fit_result.success}")
```
gives

```
TensorFlow

W MnPosDef non-positive diagonal element in covariance matrix[ 0 ] = -1.49124e+08
W MnPosDef non-positive diagonal element in covariance matrix[ 1 ] = -4983.71
W MnPosDef Added to diagonal of Error matrix a value 1.49124e+08
fit params: [0.9366918206214905, 0.9126002788543701]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 15.21                      │              Nfcn = 74               │
│ EDM = 1.14e+03 (Goal: 0.0002)    │              Ngrad = 1               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │       SOME Parameters at limit       │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │     1     │     7     │            │            │    0    │   10    │       │
│ 1 │ x1   │ 912.60e-3 │  0.33e-3  │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────────────┐
│    │       x0       x1 │
├────┼───────────────────┤
│ x0 │       34        0 │
│ x1 │        0 1.11e-07 │
└────┴───────────────────┘
success: True

PyTorch

fit params: [0.9731879234313965, 0.9999999403953552]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 17.09                      │              Nfcn = 56               │
│ EDM = 0.000389 (Goal: 0.0002)    │              Ngrad = 1               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │        No Parameters at limit        │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │ 973.19e-3 │  0.29e-3  │            │            │    0    │   10    │       │
│ 1 │ x1   │1000.00e-3 │  0.24e-3  │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────────────┐
│    │       x0       x1 │
├────┼───────────────────┤
│ x0 │ 8.61e-08 5.75e-08 │
│ x1 │ 5.75e-08 5.77e-08 │
└────┴───────────────────┘
success: True

JAX

WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
fit params: [0.5003563165664673, 0.9998618364334106]
┌──────────────────────────────────┬──────────────────────────────────────┐
│ FCN = 13.11                      │              Nfcn = 34               │
│ EDM = 1.23 (Goal: 0.0002)        │              Ngrad = 4               │
├───────────────┬──────────────────┼──────────────────────────────────────┤
│ Valid Minimum │ Valid Parameters │       SOME Parameters at limit       │
├───────────────┴──────────────────┼──────────────────────────────────────┤
│ Below EDM threshold (goal x 10)  │           Below call limit           │
├───────────────┬──────────────────┼───────────┬─────────────┬────────────┤
│  Covariance   │     Hesse ok     │ Accurate  │  Pos. def.  │ Not forced │
└───────────────┴──────────────────┴───────────┴─────────────┴────────────┘
┌───┬──────┬───────────┬───────────┬────────────┬────────────┬─────────┬─────────┬───────┐
│   │ Name │   Value   │ Hesse Err │ Minos Err- │ Minos Err+ │ Limit-  │ Limit+  │ Fixed │
├───┼──────┼───────────┼───────────┼────────────┼────────────┼─────────┼─────────┼───────┤
│ 0 │ x0   │     1     │     9     │            │            │    0    │   10    │       │
│ 1 │ x1   │     1     │     9     │            │            │  1e-10  │   10    │       │
└───┴──────┴───────────┴───────────┴────────────┴────────────┴─────────┴─────────┴───────┘
┌────┬───────────┐
│    │   x0   x1 │
├────┼───────────┤
│ x0 │  377 -188 │
│ x1 │ -188 94.2 │
└────┴───────────┘
success: True

```

but if you flip strategy to `2` TensorFlow is able to get the JAX result but PyTorch will fail.

Comment (alexander-held) [reply tests/test_optim.py:98]: I am surprised that the TensorFlow result says `Not forced`, I thought this referred to forcing the covariance matrix to be positive definite, which seems to happen in the messages above. `SOME Parameters at limit` is also an indication that the result (and in particular parameter uncertainties) may be unreliable. It is also very odd to see `EDM = 1.14e+03 (Goal: 0.0002)` and still `success: True` (JAX also has a very large EDM). Weirdly enough, the PyTorch status looks reasonable but the result not at all.

Comment (alexander-held) [reply tests/test_optim.py:98]: Is it normal to only get these few gradient calls? `Ngrad = 1` is surprising to me.

Comment (alexander-held) [reply tests/test_optim.py:98]: I experimented some more with numpy and without gradients: the 32b version seems unable to reach the EDM goal, and lowering the tolerance does not help. `iminuit` still reports that the EDM is below threshold, which I do not understand even when taking into account [the factor 10](https://iminuit.readthedocs.io/en/stable/reference.html#iminuit.Minuit.tol) into account.

Comment (matthewfeickert) [reply tests/test_optim.py:98]: Thanks for taking a look at this @alexander-held. I'm afraid I don't fully understand this behavior yet either, which might take some time.

**Comment by matthewfeickert** [comment]
I'm putting this back into Draft mode until `v0.6.1` so that we can better understand the behavior here and revise the tests.

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg This will obviously need some discussion, but I think this is ready for review now.

**Review by kratsg — APPROVED** [review]

**Comment by kratsg** [comment]
The 0.2 is the only thing I think would need discussion, but the rest seems... ok.

**Comment by matthewfeickert** [comment]
> The 0.2 is the only thing I think would need discussion, but the rest seems... ok.

Yeah, I don't really remember why I needed to do this, so I'll probably need to revisit this.

**Comment by matthewfeickert** [comment]
@kratsg I think @alexander-held's [comment here](https://github.com/scikit-hep/pyhf/pull/1306#discussion_r574780745) might be about as good as we're going to get

> I experimented some more with numpy and without gradients: the 32b version seems unable to reach the EDM goal, and lowering the tolerance does not help. `iminuit` still reports that the EDM is below threshold, which I do not understand even when taking [the factor 10](https://iminuit.readthedocs.io/en/stable/reference.html#iminuit.Minuit.tol) into account.

This could also just come down to how `Minuit2` is being implemented in ROOT(?), which is used by `iminuit`.

**Comment by kratsg** [comment]
LGTM then.

**Pull request merged**